File: x_b897(find_the_ways)/Python/find_the_ways.py
from sys import stdin
from math import log, ceil
while True:
    line = stdin.readline().strip('\n')
    if not line: break
    total, remove = map(int, line.split())
    if remove < total / 2: remove = total - remove
    if remove == total:
        print(1)
        continue
    digits = 0
    for n in range(remove + 1, total + 1):
        digits += log(n, 10)
    for n in range(1, total - remove + 1):
        digits -= log(n, 10)
    print(ceil(digits))

File: b897(find_the_ways)/Java/FindTheWays.java
import java.lang.Math;
import java.util.Scanner;


public class FindTheWays {
    public static void main(String[] args) {
        Scanner in = new Scanner(System.in);

        while (in.hasNext()) {
            long total  = in.nextLong();
            long remove = in.nextLong();

            /* all remove */
            if (remove == total) {
                System.out.println(1);
                continue;
            }
            /*
            * using combinations, make remove bigger:
            * C(10, 8) == C(10, 2) but less calculation
            */
            if (remove < (total >> 1)) {
                remove = total - remove;
            }
            /*
            * C(10, 8) = (10 * 9) / (1 * 2)
            * digits = ceil( log( C(10, 8))) = ceil (log10 + log9 - log1 - log2)
            */
            double digits = 0.0F;
            for (long n = remove + 1; n <= total; ++n) {
                digits += Math.log10(n) - Math.log10(total + 1 - n);
            }
            System.out.println((int) Math.ceil(digits));
        }
    }
}
